src/python/pants/goal/migrate_call_by_name.py

```diff
@@ -3,35 +3,655 @@
+
+
+# ------------------------------------------------------------------------------------------
+# Call-by-name syntax mapping
+# ------------------------------------------------------------------------------------------
+
+
+class CallByNameSyntaxMapper:
+    def __init__(self, graphs: list[RuleGraphGet]) -> None:
+        self.graphs = graphs
+
+    def map_get_to_new_syntax(
+        self, get: ast.Call, filename: PurePath, calling_func: str
+    ) -> Replacement | None:
+        """There are 4 forms the old Get() syntax can take, so we can account for each one of them
+        individually."""
+        new_source: ast.Call | None = None
+        imports: list[ast.ImportFrom] = []
+
+        graph_item = next(
+            (
+                item
+                for item in self.graphs
+                if item["filepath"] == str(filename) and item["function"] == calling_func
+            ),
+            None,
+        )
+        if not graph_item:
+            logger.warning(f"Failed to find dependencies for {filename} {calling_func}")
+            return None
+
+        get_deps = graph_item["gets"]
+
+        try:
+            if len(get.args) == 1:
+                new_source, imports = self.map_no_args_get_to_new_syntax(get, get_deps)
+            elif len(get.args) == 2 and isinstance(get.args[1], ast.Call):
+                new_source, imports = self.map_short_form_get_to_new_syntax(get, get_deps)
+            elif len(get.args) == 2 and isinstance(get.args[1], ast.Dict):
+                new_source, imports = self.map_dict_form_get_to_new_syntax(get, get_deps)
+            elif len(get.args) == 3:
+                new_source, imports = self.map_long_form_get_to_new_syntax(get, get_deps)
+            else:
+                logging.warning(f"Failed to migrate {ast.unparse(get)} due to unknown form\n")
+                return None
+        except AssertionError as e:
+            logging.warning(f"Failed to migrate {ast.unparse(get)} with assertion error {e}\n")
+            return None
+        except Exception as e:
+            logging.warning(f"Failed to migrate {ast.unparse(get)} with {e}\n")
+            return None
+
+        assert get.end_lineno is not None
+        assert get.end_col_offset is not None
+        return Replacement(
+            filename=filename,
+            module=graph_item["module"],
+            line_range=(get.lineno, get.end_lineno),
+            col_range=(get.col_offset, get.end_col_offset),
+            current_source=get,
+            new_source=new_source,
+            additional_imports=[
+                ast.ImportFrom(
+                    module="pants.engine.rules", names=[ast.alias("implicitly")], level=0
+                ),
+                *imports,
+            ],
+        )
+
+    def map_no_args_get_to_new_syntax(
+        self, get: ast.Call, deps: list[RuleGraphGetDep]
+    ) -> tuple[ast.Call, list[ast.ImportFrom]]:
+        """Get(<OutputType>) -> the_rule_to_call(**implicitly())"""
```

Comment:
FWIW: If the called rule doesn't take any arguments, you can skip the `**implicitly()`, and just call it as `the_rule_to_call()`.

Reply:
👍🏽 

Ah, okay, I didn't know if there were any ambient arguments that could still be passed without an explicit Get arg.

Reply:
...oh. Hm. Good question. 😬